File: src/helpers/langchain.ts
/**
 * Helper to convert a PromptConfig to LangChain.js compatible objects.
 */

import type { PromptConfig } from '../types.js';

/**
 * Prompt template configuration for LangChain.
 */
export interface LangChainPromptConfig {
  /** System message template */
  systemTemplate: string;
  /** User message template (with {variable} format) */
  userTemplate: string;
  /** List of input variable names */
  inputVariables: string[];
}

/**
 * Model configuration for LangChain.
 */
export interface LangChainModelConfig {
  /** Model name/ID */
  modelName: string;
  /** Temperature setting */
  temperature: number;
  /** Max output tokens */
  maxTokens: number;
  /** Top P sampling */
  topP?: number;
  /** Top K sampling */
  topK?: number;
  /** Frequency penalty */
  frequencyPenalty?: number;
  /** Presence penalty */
  presencePenalty?: number;
  /** Stop sequences */
  stop?: string[];
}

/**
 * Extract input variable names from a template string.
 * Converts {{variable}} format to LangChain's {variable} format.
 */
function extractVariables(template: string | undefined): string[] {
  const matches = template?.match(/\{\{(\w+)\}\}/g) || [];
  return [...new Set(matches.map((m) => m.slice(2, -2)))];
}

/**
 * Convert {{variable}} placeholders to LangChain's {variable} format.
 */
function convertTemplate(template: string | undefined): string {
  if (!template) return '';
  return template.replace(/\{\{(\w+)\}\}/g, '{$1}');
}

/**
 * Convert a PromptConfig to LangChain prompt template configuration.
 *
 * @example
 * ```ts
 * import { toPromptTemplate } from 'lit-prompt-config/helpers/langchain';
 * import { ChatPromptTemplate } from '@langchain/core/prompts';
 *
 * const templateConfig = toPromptTemplate(config);
 *
 * const prompt = ChatPromptTemplate.fromMessages([
 *   ['system', templateConfig.systemTemplate],
 *   ['user', templateConfig.userTemplate],
 * ]);
 * ```
 */
export function toPromptTemplate(config: PromptConfig): LangChainPromptConfig {
  const userVars = extractVariables(config.userPromptTemplate);

  return {
    systemTemplate: config.systemPrompt || '',
    userTemplate: convertTemplate(config.userPromptTemplate),
    inputVariables: userVars,
  };
}

/**
 * Convert a PromptConfig to LangChain model configuration.
 *
 * @example
 * ```ts
 * import { getModelConfig } from 'lit-prompt-config/helpers/langchain';
 * import { ChatOpenAI } from '@langchain/openai';
 *
 * const modelConfig = getModelConfig(config);
 *
 * // For OpenRouter, use with base URL
 * const llm = new ChatOpenAI({
 *   ...modelConfig,
 *   configuration: {
 *     baseURL: 'https://openrouter.ai/api/v1',
 *   },
 * });
 * ```
 */
export function getModelConfig(config: PromptConfig): LangChainModelConfig {
  const result: LangChainModelConfig = {
    modelName: config.model,
    temperature: config.temperature,
    maxTokens: config.maxTokens,
  };

  if (config.topP != null && config.topP !== 1) {
    result.topP = config.topP;
  }
  if (config.topK) {
    result.topK = config.topK;
  }
  if (config.frequencyPenalty) {
    result.frequencyPenalty = config.frequencyPenalty;
  }
  if (config.presencePenalty) {
    result.presencePenalty = config.presencePenalty;
  }
  if (config.stopSequences?.length) {
    result.stop = config.stopSequences;
  }

  return result;
}

/**
 * Get the provider name from an OpenRouter model ID.
 *
 * @example
 * ```ts
 * getProvider('anthropic/claude-sonnet-4-5') // 'anthropic'
 * getProvider('openai/gpt-4o') // 'openai'
 * ```
 */
export function getProvider(modelId: string): string {
  return modelId.split('/')[0] || 'unknown';
}

File: helpers/langchain.py
"""
lit-prompt-config LangChain integration helper.

Convert lit-prompt-config JSON to LangChain prompt templates and LLM instances.

Usage:
    from lit_prompt_config import from_config

    config = load_your_config_json()  # e.g., from database, file, etc.
    prompt, llm = from_config(config)

    chain = prompt | llm
    result = chain.invoke({"article_text": "..."})
"""

import re
from typing import Any, Tuple, List, Optional


def extract_variables(template: Optional[str]) -> List[str]:
    """Extract variable names from a template with {{variable}} placeholders."""
    if not template:
        return []
    matches = re.findall(r'\{\{(\w+)\}\}', template)
    return list(set(matches))


def convert_template(template: Optional[str]) -> str:
    """Convert {{variable}} placeholders to LangChain's {variable} format."""
    if not template:
        return ""
    return re.sub(r'\{\{(\w+)\}\}', r'{\1}', template)


def to_prompt_template(config: dict) -> "ChatPromptTemplate":
    """
    Convert a PromptConfig to a LangChain ChatPromptTemplate.

    Args:
        config: The prompt configuration dictionary

    Returns:
        A ChatPromptTemplate instance

    Example:
        >>> prompt = to_prompt_template(config)
        >>> prompt.invoke({"article_text": "..."})
    """
    from langchain_core.prompts import ChatPromptTemplate

    messages = []

    if config.get('systemPrompt'):
        messages.append(('system', config['systemPrompt']))

    user_template = convert_template(config.get('userPromptTemplate', ''))
    if user_template:
        messages.append(('user', user_template))

    return ChatPromptTemplate.from_messages(messages)


def get_model_kwargs(config: dict) -> dict:
    """
    Extract model configuration kwargs from a PromptConfig.

    Args:
        config: The prompt configuration dictionary

    Returns:
        Dictionary of model kwargs (temperature, max_tokens, etc.)
    """
    kwargs = {
        'temperature': config.get('temperature', 1.0),
        'max_tokens': config.get('maxTokens', 4096),
    }

    if config.get('topP') is not None and config.get('topP') != 1:
        kwargs['top_p'] = config['topP']

    if config.get('stopSequences'):
        kwargs['stop'] = config['stopSequences']

    return kwargs


def get_provider(model_id: str) -> str:
    """
    Extract provider name from an OpenRouter model ID.

    Args:
        model_id: e.g., "anthropic/claude-sonnet-4-5"

    Returns:
        Provider name, e.g., "anthropic"
    """
    return model_id.split('/')[0] if '/' in model_id else 'unknown'


def create_llm(config: dict, api_key: Optional[str] = None) -> Any:
    """
    Create a LangChain LLM instance from a PromptConfig.

    Supports Anthropic, OpenAI, and falls back to OpenRouter via OpenAI-compatible API.

    Args:
        config: The prompt configuration dictionary
        api_key: Optional API key (uses environment variables if not provided)

    Returns:
        A LangChain chat model instance

    Example:
        >>> llm = create_llm(config, api_key="sk-...")
        >>> llm.invoke([HumanMessage(content="Hello")])
    """
    model_id = config.get('model', '')
    provider = get_provider(model_id)
    kwargs = get_model_kwargs(config)

    if provider == 'anthropic':
        from langchain_anthropic import ChatAnthropic
        # Extract the model name without provider prefix
        model_name = model_id.split('/')[-1] if '/' in model_id else model_id
        if api_key:
            kwargs['anthropic_api_key'] = api_key
        return ChatAnthropic(model=model_name, **kwargs)

    elif provider == 'openai':
        from langchain_openai import ChatOpenAI
        model_name = model_id.split('/')[-1] if '/' in model_id else model_id
        if api_key:
            kwargs['openai_api_key'] = api_key
        return ChatOpenAI(model=model_name, **kwargs)

    else:
        # Default to OpenRouter via OpenAI-compatible API
        from langchain_openai import ChatOpenAI
        return ChatOpenAI(
            model=model_id,
            base_url='https://openrouter.ai/api/v1',
            api_key=api_key,
            **kwargs
        )


def from_config(
    config: dict,
    api_key: Optional[str] = None
) -> Tuple["ChatPromptTemplate", Any]:
    """
    Convert a lit-prompt-config JSON to LangChain prompt template and LLM.

    This is the main entry point for using lit-prompt-config with LangChain.

    Args:
        config: The prompt configuration dictionary (from the <prompt-config> component)
        api_key: Optional API key for the LLM provider

    Returns:
        Tuple of (ChatPromptTemplate, ChatModel)

    Example:
        >>> import json
        >>>
        >>> # Load config from wherever you stored it
        >>> with open('my-prompt.json') as f:
        ...     config = json.load(f)
        >>>
        >>> # Create LangChain objects
        >>> prompt, llm = from_config(config, api_key="sk-...")
        >>>
        >>> # Use in a chain
        >>> chain = prompt | llm
        >>> result = chain.invoke({"article_text": "The quick brown fox..."})
        >>> print(result.content)
    """
    prompt = to_prompt_template(config)
    llm = create_llm(config, api_key)
    return prompt, llm


# For backwards compatibility
__all__ = [
    'from_config',
    'to_prompt_template',
    'create_llm',
    'get_model_kwargs',
    'get_provider',
    'extract_variables',
    'convert_template',
]
